docs/types.rst

```diff
@@ -914,6 +921,54 @@ Of course, you can also directly access the members of the struct without
 assigning it to a local variable, as in
 ``campaigns[campaignID].amount = 0``.
 
+
+Types as input and output parameters
+====================================
+
+.. index:: !function parameters, functions, parameters
+
+Allowed function parameter types
+--------------------------------
+
+**All** :ref:`value-types` variables are allowed as input and output function parameter types.
+
+**All** :ref:`reference-types` variables are allowed as input and output function parameter types for internal functions, and for external functions if you enable the experimental ``ABIEncoderV2`` feature.
+
+.. index:: copying types, referencing types
+
+.. _ref-copy-types:
+
+Copying vs referencing
+----------------------
+
+When you use variables as an input or output parameter to functions, how the
+compiler treats the variable depends on its type.
+
+Any variable that is a :ref:`value-types`, has its value **copied** from the
+output function to the input function.
+
+For variables that are more complex types (or :ref:`reference-types`), whether
+the variable is copied to or referenced to the input function follows the same
+rules as :ref:`data-location-assignment`, in summary:
+
+- Variables appended with ``storage`` in the output function and ``memory`` in the input function have their values **copied** between the functions.
+- Variables appended with ``storage`` in the output function and ``storage`` in the input function have their values **referenced** between the functions.
+- Variables already stored in ``memory`` in the output function and variables already stored in ``memory`` in the input function have their values **referenced** between the functions.
+
+When a complex variable is copied from an output function to an input function
```

Comment:
Perhaps it would be better to single this out into its own section because it applies both to function calls and copies in assignments.